scionlab/tests/test_user_as_models.py

```diff
@@ -213,37 +240,83 @@ def _check_attachment_point(testcase, attachment_point):
     only seems to be used here.
     """
-    prev_ap = user_as.attachment_point
-
     with patch.object(AttachmentPoint, 'trigger_deployment', autospec=True) as mock_deploy:
         user_as.update(
-            attachment_point=kwargs.get('attachment_point', user_as.attachment_point),
             label=kwargs.get('label', user_as.label),
             installation_type=kwargs.get('installation_type', user_as.installation_type),
-            use_vpn=kwargs.get('use_vpn', user_as.is_use_vpn()),
-            public_ip=kwargs.get('public_ip', user_as.public_ip),
-            public_port=kwargs.get('public_port', user_as.get_public_port()),
-            bind_ip=kwargs.get('bind_ip', user_as.bind_ip),
-            bind_port=kwargs.get('bind_port', user_as.bind_port),
         )
+        user_as.update_attachments(aps_confs)
 
-    # Check that deployment was triggered once for each of the attachment points.
-    curr_ap = user_as.attachment_point
+    # Check that deployment was triggered strictly once for each attachment point
     testcase.assertEqual(
-        sorted((args[0] for args, kwargs in mock_deploy.call_args_list), key=id),
-        sorted(set([prev_ap, curr_ap]), key=id)
+        len([args[0] for args, kwargs in mock_deploy.call_args_list]),
+        len(set(args[0] for args, kwargs in mock_deploy.call_args_list))
+            )
+    # Check that deployment was triggered for all the attachment points
+    testcase.assertEqual(
+        set(args[0] for args, kwargs in mock_deploy.call_args_list),
+        set(AttachmentConf.attachment_points(aps_confs)),
     )
 
 
-def _get_random_useras_params(seed, force_public_ip=False, force_bind_ip=False, **kwargs):
+def _get_random_aps_params(seed,
+                           attachment_points: List[AttachmentPoint],
+                           vpn_choice: VPNChoice,
+                           force_public_ip=False,
+                           force_bind_ip=False) -> List[AttachmentConf]:
+    """
+    Generate random compatible parameters for the given AttachmentPoints based on `seed`.
+    """
+    r = random.Random(seed)
+    aps_confs = []
+    used_public_ip_port_pairs = set()
+    used_bind_ip_port_pairs = set()
+    for ap in attachment_points:
+        ap_conf_dict = {}
+        ap_conf_dict['attachment_point'] = ap
+        if vpn_choice in (VPNChoice.NONE, VPNChoice.ALL):
+            ap_conf_dict['use_vpn'] = False if vpn_choice is VPNChoice.NONE else True
+        else:
+            if not ap.vpn:
+                ap_conf_dict['use_vpn'] = False
+            else:
+                ap_conf_dict['use_vpn'] = _randbool(r)
+        while True:
+            public_ip = '172.31.0.%i' % r.randint(10, 254)
+            public_port = r.choice(range(DEFAULT_PUBLIC_PORT, DEFAULT_PUBLIC_PORT + 20))
+            if (public_ip, public_port) not in used_public_ip_port_pairs:
+                break
+        if _randbool(r) or vpn_choice is not VPNChoice.ALL or force_public_ip:
```

Comment:
The check should be on whether vpn is enabled for this config or not (i.e. `ap_conf_dict['use_vpn']`, but probably put this into a variable).

Reply:
Maybe I didn't get what you mean, but you're suggesting to make a check against `ap_conf_dict['use_vpn']`?
Here we are building the dictionary, so there's nothing in it ...